### DISCUSSION TITLE
Is Windows App Runtime backward compatible

### DISCUSSION THREAD

**Discussion by harshith187** [microsoft/WindowsAppSDK#3657] [Q&A] [answered]
Hi,

I am working on a WinUI3 application for the desktop application in cpp. I am trying to build the application as a single executable(.exe). So I chose framework dependant unpackaged mode. When we chose a framework dependant unpackaged mode of development for WinUI3 we should [explicitly install Windows App Runtime](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/deploy-unpackaged-apps#overview) in the end user's machine.

I installed the latest Windows App Runtime (version 1.3) on my machine and then tried to run an App that is built for Windows App Runtime (version 1.2), but I got this error saying

![image](https://github.com/microsoft/WindowsAppSDK/assets/29576770/8345cc56-ad2a-4744-a0fe-ba09a243dfd9)
I have the latest Windows app runtime installed

![image](https://github.com/microsoft/WindowsAppSDK/assets/29576770/d1b1835c-fd64-49d3-b8d3-be7924770106)

Does this mean Windows App Runtime is not backward compatible? Or is there any whay to overcome this drawback?

Thank you

**Comment by DarranRowe** [comment, marked as answer]
This would require manually using the bootstrapper API. You can't use the auto initialiser (i.e. you can't use the WindowsPackageType set to None) to do this since this always searches for the version of the Windows App SDK it was shipped with.
If you look at the definition for the [MddBootstrapInitialize2](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/win32/mddbootstrap/nf-mddbootstrap-mddbootstrapinitialize2) function, you should notice that the first parameter is majorMinorVersion. This allows you to manually select the version of the Windows App SDK that you want to load. This means, if you call it like:
```
HRESULT hr = MddBootstrapInitialize2((1u) << 16 | 2u, L"", {}, MddBootstrapInitializeOptions_None);
```
you can select any version of the Windows App SDK.

If you are using a version of the Windows App SDK that the project wasn't built against, you should perform extra checks. Going forward should be safer, but going backwards is dangerous. It means that you have to take care when using functionality that was introduced in newer versions of the Windows App SDK. C++/WinRT will just use the functionality of the version that the headers were generated against.

**Reply by harshith187** [reply]
Thank you this was very helpful. Had one doubt, suppose my app is built for Windows App SDK version 1.2 and my machine only has Windows App SDK 1.3 installed. How will I know from c++ that Windows app SDK 1.3 is installed on the machine so that I can bootstrap initialize to that version?

**Reply by DarranRowe** [reply]
You would use the Windows [package manager](https://learn.microsoft.com/en-us/uwp/api/windows.management.deployment.packagemanager?view=winrt-22621) to look for it. [PackageManager.FindPackagesForUser](https://learn.microsoft.com/en-us/uwp/api/windows.management.deployment.packagemanager.findpackagesforuser?view=winrt-22621) in particular allows you to look for the presence of the package for the current user if you use an empty string for the user. The two parameter version allows you to look for the package family name and the three parameter version allows you to look for the package name and publisher.
If you remember that the package name is always like `Microsoft.WindowsAppRuntime.1.3`, and the package family name is always like `Microsoft.WindowsAppRuntime.1.3_8wekyb3d8bbwe`, which is the package name with Microsoft's publisher ID on the end, then you can build a versioned string for any version.

**Reply by DrusTheAxe** [reply]
> This would require manually using the bootstrapper API. You can't use the auto initialiser

Correct. The auto-initializer will use the version of Windows App SDK that you're built with.

However, if you build with 1.2 then you probably don't want to run with 1.3. You want to run with the major.minor version of WinAppSDK that you build with. 

It's possible to build code built with one version and run with a another version - a component library is a good example why you'd want to go there (build with 1.2 but run with 1.2, 1.3, 1.4, ...) but then you're just a component in someone else's process. You have no control over process level state and have to adapt to whatever the app's chosen to do, so you certainly wouldn't be using the Bootstrapper API or otherwise deciding the version of WinAppSDK in use by the app. But if you're building an app build-time version = run-time version is generally expected.

Hmmm. WinAppSDK's versioning rules are well established but I can't find a link where it's clearly documented. Closest I'm aware of is
[2.2. Decision 2: Breaking Change Boundary for Version 1.x](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/Deployment/MSIXPackageVersioning.md#22-decision-2-breaking-change-boundary-for-version-1x) but that's only partially related. @bpulliam got link to where WinAppSDK's versioning model is defined?

**Reply by harshith187** [reply]
Thank you this was very helpful
